### PR TITLE
Add GoReleaser

### PR DESCRIPTION
**Pull request by lucacome**
~This will get the wrong `tag` until #1544 is merged.~

Removed `edge` from the version name in the Makefile, it will now be something like `v1.11.1-SNAPSHOT-00f0053` for local builds (same that GoReleaser uses for `edge`), so it will remove the need to update the version in the Makefile.
(When we make a stable release, GoReleaser will use `v1.11.1`)

GoReleaser builds binaries for:
- amd64
- arm
- arm64
- ppc64le
- s390x

and we push Docker images for the same architectures.

Addresses https://github.com/nginxinc/kubernetes-ingress/issues/944, https://github.com/nginxinc/kubernetes-ingress/issues/974 and https://github.com/nginxinc/kubernetes-ingress/issues/1310

### CODE REVIEW

File: .github/workflows/edge.yml [hunk 1]
```diff
@@ -46,6 +45,8 @@ jobs:
     steps:
       - name: Checkout Repository
         uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
```
Comment (soneillf5) [inline .github/workflows/edge.yml:49]: why do we do this? does it break otherwise ?

File: .gitignore [hunk 2]
```diff
@@ -53,3 +53,5 @@ hack/controller-gen-*
 docs-web/.netlify/state.json
 site/
 venv/
+
+dist/
```
Comment (soneillf5) [inline .gitignore:57]: cool!

**Review by soneillf5 — APPROVED** [review]
Really cool! It's so different from the stuff I learnt when starting go about targeting all the different architectures.

File: .github/workflows/edge.yml (hunk 1 above)
Comment (lucacome) [reply .github/workflows/edge.yml:49]: yeah, the docs for GoReleaser say:

> Note the fetch-depth: 0 option on the Checkout workflow step. It is required for GoReleaser to work properly. Without that, GoReleaser might fail or behave incorrectly.

this is the description of the parameter:

>Number of commits to fetch. 0 indicates all history for all branches and tags.

File: .goreleaser.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+env:
+  - GO111MODULE=on
+before:
+  hooks:
+    - go mod tidy
```
Comment (pleshakov) [inline .goreleaser.yml:5]: are `tidy` and `verify` commands really necessary?
if they are expected to change the repo contents, would it be better if our pipeline had a check in a separate stage - to check that the repo files do not change after tidy and verify?

File: build/Dockerfile [hunk 4]
```diff
@@ -272,7 +272,7 @@ RUN mkdir -p /var/lib/nginx /etc/nginx/secrets /etc/nginx/stream-conf.d \
 	&& chown -R nginx:0 /etc/nginx /var/cache/nginx /var/lib/nginx \
 	&& rm -f /etc/nginx/conf.d/* /etc/apt/apt.conf.d/90nginx /etc/apt/sources.list.d/nginx-plus.list
 
-COPY internal/configs/version1/nginx$PLUS.ingress.tmpl \
+COPY --chown=nginx:0 internal/configs/version1/nginx$PLUS.ingress.tmpl \
```
Comment (pleshakov) [inline build/Dockerfile:275]: why is this change necessary? 
for the  template files, it is ok if root owns it, because the IC only need to read it and only to read it on the start.

File: build/Dockerfile [hunk 5]
```diff
@@ -284,28 +284,36 @@ COPY internal/configs/version1/nginx$PLUS.ingress.tmpl \
...
 COPY --chown=nginx:0 nginx-ingress /
 
-USER nginx
+
+############################################# Create image with nginx-ingress built by GoReleaser #############################################
+FROM common AS goreleaser
+ARG TARGETARCH
+ARG TARGETVARIANT
+
+COPY --chown=nginx:0 dist/kubernetes-ingress_linux_$TARGETARCH${TARGETVARIANT:+_7}/nginx-ingress /
```
Comment (pleshakov) [inline build/Dockerfile:319]: similar question, why does it need to be owned by nginx user?

File: cmd/nginx-ingress/main.go [hunk 6]
```diff
@@ -195,7 +196,7 @@ func main() {
 	}
 
 	if *versionFlag {
-		fmt.Printf("Version=%v GitCommit=%v\n", version, gitCommit)
+		fmt.Printf("Version=%v GitCommit=%v Date=%v\n", version, commit, date)
```
Comment (pleshakov) [inline cmd/nginx-ingress/main.go:199]: do we need to update the doc for the version flag?

Comment (pleshakov) [inline cmd/nginx-ingress/main.go:199]: is it clear from the context that it is a build date? Perhaps `BuildDate` ?

File: .goreleaser.yml (hunk 3 above)
Comment (lucacome) [reply .goreleaser.yml:5]: they're not expected to change the repository's content, it's just a check to make sure everything is fine.

File: build/Dockerfile (hunk 4 above)
Comment (lucacome) [reply build/Dockerfile:275]: We had a problem not long ago with permissions and having nginx be the owner of the files will fix that. Since we set nginx as the user and run the binary as nginx user, it makes sense that nginx owns the files.

File: build/Dockerfile (hunk 5 above)
Comment (lucacome) [reply build/Dockerfile:319]: it's the same for the binaries in the other layers.

File: cmd/nginx-ingress/main.go (hunk 6 above)
Comment (lucacome) [reply cmd/nginx-ingress/main.go]: I've updated the flag description, is that it?

Comment (lucacome) [reply cmd/nginx-ingress/main.go]: I think it's clear from the context, like for the git commit, but I'm fine either way.

File: .goreleaser.yml (hunk 3 above)
Comment (pleshakov) [reply .goreleaser.yml:5]: `go mod tidy` exits with 0 status code even if it updates the go.mod. so it will silently fix the files.

File: build/Dockerfile (hunk 4 above)
Comment (pleshakov) [reply build/Dockerfile:275]: for the files that the IC or NGINX workers can change, it makes sense. But for the files that IC or NGINX only need to read, I think it makes sense to be owned by root and allow read/execute on those, to follow the principle of least privilege. For example, as NGINX binary:
```
ls -la /usr/sbin/nginx
-rwxr-xr-x 1 root root 1698544 Nov 26 13:00 /usr/sbin/nginx
```

File: cmd/nginx-ingress/main.go (hunk 6 above)
Comment (pleshakov) [reply cmd/nginx-ingress/main.go]: looks good. thanks!
we have the flag docs here as well though https://docs.nginx.com/nginx-ingress-controller/configuration/global-configuration/command-line-arguments/

**Review by pleshakov — APPROVED** [review]

File: build/Dockerfile (hunk 4 above)
Comment (lucacome) [reply build/Dockerfile:275]: isn't it least privilege if the user is nginx instead of root?
